### ISSUE TITLE
[ERROR] Unable to connect to psql running inside a docker container

### ISSUE THREAD

**Issue by sajjathossain** [tpope/vim-dadbod#142]
@tpope I have a Postgresql database running inside a docker container. I'm unable to connect to that database. getting psql executable not found error.

Here's my configuration,
```lua
return {
  "tpope/vim-dadbod",
  dependencies = {
    "kristijanhusak/vim-dadbod-ui",
    "kristijanhusak/vim-dadbod-completion",
  },
  config = function()
    require("plugins.configs.dadbod.config").setup()
  end,
  cmd = { "DBUIToggle", "DBUI", "DBUIAddConnection", "DBUIFindBuffer", "DBUIRenameBuffer", "DBUILastQueryInfo" },
}
```
config file,
```lua
local M = {}

-- local function db_completion()
--   require("cmp").setup.buffer { sources = { { name = "vim-dadbod-completion" } } }
-- end

function M.setup()
  vim.g.db_ui_save_location = vim.fn.stdpath "plugins.configs.dadbod" .. require("plenary.path").path.sep .. "db_ui"

  vim.api.nvim_create_autocmd("FileType", {
    pattern = {
      "sql",
    },
    command = [[setlocal omnifunc=vim_dadbod_completion#omni]],
  })

  vim.api.nvim_create_autocmd("FileType", {
    pattern = {
      "sql",
      "mysql",
      "psql",
    },
    callback = function()
      vim.schedule()
    end,
  })
end

return M
```

**Comment by tpope** [comment]
Do you have `psql` installed? Sounds like you need to install it.

**Comment by TimothyGCY** [comment]
I have the same issue using MySQL, trying to connect to DB on Docker, exposing default port. If I am to install `mysql` on my machine, the service will get conflicted.

**Comment by tpope** [comment]
You only need the client, not the server. Or you can install both and disable the server.

**Comment by lpanebr** [comment]
> You only need the client, not the server. Or you can install both and disable the server.

I was also having trouble with a  `docker container` (but `mysql`)...

Installing just the `mysql-client` as instructed by @tpope *almost worked for me*.

**I then got the error below:**

    ERROR 2002 (HY000): Can't connect to local MySQL server through socket '/var/run/mysqld/mysqld.sock' (2)

**which was resolved by** using `127.0.0.1` instead of `localhost` in the connection string.

*The weird thing is that my /etc/hosts already had the mapping:*
```zsh
❯ cat /etc/hosts
127.0.0.1       localhost
```

*Perhaps the plugin could do some magic to use IP when localhost is provided.. but I have no idea if this would the better or worse for most people...* 🤨

I hope this helps!

**Comment by kareem-abdul** [comment]
So I have been trying to make it work with docker without installing any client software and found a way to achieve this.

What I did was
 - I created a script named the same as the client I'm trying to support
 - added the following docker command inside the script
    `docker exec -i <container_name> <client_name> $@` 
    The `-i` flag is important as it creates an open stdin that does not close till the container stops.
- made this script available in my PATH. 
  Now when I try to connect to a db vim-dadbod's adapter for that db will execute this script.

For example, to support MongoDB, I created a simple script called `mongosh` as follows:
```sh
#!/bin/bash

docker exec -i mongodb_container mongosh $@
```
I tested the same with mysql, psql, redis. 
This might work on other adapters too, but might need to test it and confirm.

Also, the above method works so well for me, that I don't think adding support for docker inside adapters is needed.

One issue with this script is that the container name/ID might need to change frequently based on how you create containers. 
Since I always use Tmux, changing the container name/ID is as simple as performing an fzf search inside a Tmux prompt.
A much simpler approach would be to read the container ID from an environment variable.

**Comment by greyHairChooseLife** [comment]
@kareem-abdul 
Thanks for your help messages.

Would you please explain further about what you said like below, if you don't mind? 
It's just out of my understanding.

>made this script available in my PATH.
Now when I try to connect to a db vim-dadbod's adapter for that db will execute this script.

**Comment by kareem-abdul** [comment]
> @kareem-abdul Thanks for your help messages.
> 
> Would you please explain further about what you said like below, if you don't mind? It's just out of my understanding.
> 
> > made this script available in my PATH.
> > Now when I try to connect to a db vim-dadbod's adapter for that db will execute this script.

@greyHairChooseLife  sure,
Vim-dadbod has adapters for each database/client, which generate commands for executing queries and making connections. For instance, the MongoDB adapter uses the `mongosh` or `mongo` client. You can see this in the adapter code 
https://github.com/tpope/vim-dadbod/blob/7888cb7164d69783d3dce4e0283decd26b82538b/autoload/db/adapter/mongodb.vim#L1-L5

When you run vim-dadbod commands like `:DB mongodb:///test`, it tries to execute `mongosh` or `mongo` command. This is defined in the [db.vim file](https://github.com/tpope/vim-dadbod/blob/master/autoload/db.vim#L401)
Typically, you'd install the MongoDB client to get these commands, but without it, you'll encounter errors.

Instead of installing the MongoDB client, I created a `mongosh` bash script and added it to my `PATH`. This way, when vim-dadbod looks for `mongosh`, it finds my script and executes it with the arguments generated by the adapter. 

You can find where your PATH locations for a linux system are by executing `echo $PATH` in a terminal.

Basically, the script in my previous comment acts as a delegator, which passes on whatever vim-dadbod wants to execute to the mongosh command inside a container.
